crates/ra_ide/src/ssr.rs

```diff
@@ -0,0 +1,275 @@
+    }
+}
+
+fn is_expr(s: &str) -> Result<(), SsrError> {
+    if s == "expr" {
+        Ok(())
+    } else {
+        Err(SsrError::ParseError("Only $<name>:expr is supported".into()))
+    }
+}
+
+fn replace_in_template(template: String, name: &str, new_name: &str) -> String {
+    let name = format!("${}", name);
+    template.replace(&name, new_name)
+}
+
+fn create_name<'a>(name: &str, binding_names: &'a mut Vec<String>) -> Result<&'a str, SsrError> {
+    let sanitized_name = format!("__search_pattern_{}", name);
+    if binding_names.iter().any(|a| a == &sanitized_name) {
+        return Err(SsrError::ParseError(format!("Name `{}` repeats more than once", name)));
+    }
+    binding_names.push(sanitized_name);
+    Ok(binding_names.last().unwrap())
+}
+
+pub(crate) fn find(pattern: &SsrPattern, code: &SyntaxNode) -> SsrMatches {
+    fn check(
+        pattern: &SyntaxElement,
+        code: &SyntaxElement,
+        placeholders: &[String],
+        m: &mut Match,
+    ) -> bool {
+        match (pattern, code) {
+            (SyntaxElement::Token(ref pattern), SyntaxElement::Token(ref code)) => {
+                pattern.text() == code.text()
+            }
+            (SyntaxElement::Node(ref pattern), SyntaxElement::Node(ref code)) => {
+                if placeholders.iter().find(|&n| n.as_str() == pattern.text()).is_some() {
+                    m.binding.insert(pattern.text().to_string(), code.clone());
+                    true
+                } else {
+                    pattern.green().children().count() == code.green().children().count()
+                        && pattern
+                            .children_with_tokens()
+                            .zip(code.children_with_tokens())
+                            .all(|(a, b)| check(&a, &b, placeholders, m))
+                }
+            }
+            _ => false,
+        }
+    }
+    let kind = pattern.pattern.kind();
+    let matches = code
+        .descendants_with_tokens()
+        .filter(|n| n.kind() == kind)
+        .filter_map(|code| {
+            let mut m = Match { place: code.as_node().unwrap().clone(), binding: HashMap::new() };
+            if check(&SyntaxElement::from(pattern.pattern.clone()), &code, &pattern.names, &mut m) {
+                Some(m)
+            } else {
+                None
+            }
+        })
+        .collect();
+    SsrMatches { matches }
+}
+
+pub(crate) fn replace(matches: &SsrMatches, template: &SsrTemplate) -> TextEdit {
+    let mut builder = TextEditBuilder::default();
+    for m in &matches.matches {
+        builder.replace(m.place.text_range(), render(&m.binding, template));
+    }
+    builder.finish()
+}
+
+fn render(binding: &Binding, template: &SsrTemplate) -> String {
+    fn replace(p: &SyntaxNode, binding: &Binding, builder: &mut TextEditBuilder) {
+        if let Some(name) = binding.keys().find(|&n| n.as_str() == p.text()) {
```

Comment:
I think it would be clearer and perhaps more efficient to pre-process `SsrTemplate` like this:

```
struct SsrTemplates {
    node: SyntaxNode,
    placeholders: FxHashMap<SyntaxElement, Var>,
}
```

we can then do roughly the following:

```rust
for element in template.node.descendats_with_tokens() {
    if let Some(var) = template.placeholders.get(element) {
        builder.replace(element.text_range(), binding[var])
    }
}
```